// @microsoft/mezzurite-core/esm5/src/utils/performance-global.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Global Mezzurite object
 */
var /**
 * Global Mezzurite object
 */
MezzuriteObject = /** @class */ (function () {
    function MezzuriteObject() {
        this.firstViewLoaded = false;
        this.captureCycleStarted = false;
        this.routerPerf = false;
        this.measures = [];
        this.defaultLogs = [];
        this.childElementNames = {};
        this.slowestResource = {};
        this.currentComponents = {};
        this.vltComponentLookup = {};
        this.elementLookup = {};
    }
    return MezzuriteObject;
}());
/**
 * Global Mezzurite object
 */
export { MezzuriteObject };
if (false) {
    /** @type {?} */
    MezzuriteObject.prototype.firstViewLoaded;
    /** @type {?} */
    MezzuriteObject.prototype.captureCycleStarted;
    /** @type {?} */
    MezzuriteObject.prototype.routerPerf;
    /** @type {?} */
    MezzuriteObject.prototype.measures;
    /** @type {?} */
    MezzuriteObject.prototype.defaultLogs;
    /** @type {?} */
    MezzuriteObject.prototype.childElementNames;
    /** @type {?} */
    MezzuriteObject.prototype.slowestResource;
    /** @type {?} */
    MezzuriteObject.prototype.currentComponents;
    /** @type {?} */
    MezzuriteObject.prototype.vltComponentLookup;
    /** @type {?} */
    MezzuriteObject.prototype.elementLookup;
}
