// @microsoft/mezzurite-core/esm5/src/services/performance-timing.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { MezzuriteConstants } from '../utils/performance-constants';
import { MezzuriteUtils } from './performance-utils.service';
import { MezzuriteMeasure } from '../utils/performance-measure';
/**
 * Class containing core timing functions
 */
var /**
 * Class containing core timing functions
 */
PerformanceTimingService = /** @class */ (function () {
    function PerformanceTimingService() {
    }
    /**
     * Creates measure object from given set of performance marks
     * @param name full semicolon delimited name
     * @param slowestResource Slowest resource inside component
     * @param maxComponentEndTime Max component end time (if VLT)
     */
    /**
     * Creates measure object from given set of performance marks
     * @param {?} name full semicolon delimited name
     * @param {?=} slowestResource Slowest resource inside component
     * @param {?=} maxComponentEndTime Max component end time (if VLT)
     * @return {?}
     */
    PerformanceTimingService.measure = /**
     * Creates measure object from given set of performance marks
     * @param {?} name full semicolon delimited name
     * @param {?=} slowestResource Slowest resource inside component
     * @param {?=} maxComponentEndTime Max component end time (if VLT)
     * @return {?}
     */
    function (name, slowestResource, maxComponentEndTime) {
        if (slowestResource === void 0) { slowestResource = null; }
        if (maxComponentEndTime === void 0) { maxComponentEndTime = null; }
        /** @type {?} */
        var startEntry;
        /** @type {?} */
        var endEntry;
        /** @type {?} */
        var componentTitle = MezzuriteUtils.getFullNamePart(name, MezzuriteConstants.fullNamePartTitle);
        /** @type {?} */
        var key = MezzuriteUtils.getFullNamePart(name, MezzuriteConstants.fullNamePartKey);
        if (name === undefined) {
            return;
        }
        if (componentTitle === MezzuriteConstants.altName) {
            // ALT
            endEntry = performance.getEntriesByName(MezzuriteConstants.altMarkEnd)[0];
        }
        else if (componentTitle === MezzuriteConstants.vltName) {
            // VLT
            startEntry = performance.getEntriesByName(MezzuriteConstants.vltMarkStart)[0];
            endEntry = {
                startTime: maxComponentEndTime
            };
        }
        else {
            // Component
            startEntry = performance.getEntriesByName(key + MezzuriteConstants.componentMarkStart)[0];
            endEntry = performance.getEntriesByName(key + MezzuriteConstants.componentMarkEnd)[0];
        }
        /** @type {?} */
        var renderStartEntry = performance.getEntriesByName(key + MezzuriteConstants.componentMarkRenderStart)[0];
        /** @type {?} */
        var startTime = startEntry !== undefined ? startEntry.startTime : 0;
        /** @type {?} */
        var endTime = endEntry.startTime;
        // reset end time to end of slowest resource if
        if (slowestResource !== null && slowestResource.responseEnd > endTime) {
            endTime = slowestResource.responseEnd;
        }
        /** @type {?} */
        var mountDuration = endEntry.startTime - startTime;
        /** @type {?} */
        var totalDuration = endTime - startTime;
        /** @type {?} */
        var nameArr = name.split(';');
        /** @type {?} */
        var obj = new MezzuriteMeasure();
        obj.name = nameArr[1];
        obj.id = nameArr[2];
        obj.startTime = startTime % 1 !== 0 ? parseFloat(startTime.toFixed(1)) : startTime;
        obj.endTime = parseFloat(endTime.toFixed(1));
        obj.untilMount = parseFloat(mountDuration.toFixed(1));
        obj.clt = parseFloat(totalDuration.toFixed(1));
        obj.slowResource = {};
        if (slowestResource && slowestResource.responseEnd >= startTime) {
            (/** @type {?} */ (obj)).slowResource['endTime'] = parseFloat(slowestResource.responseEnd.toFixed(1));
            (/** @type {?} */ (obj)).slowResource['name'] = slowestResource.name;
        }
        else if (slowestResource !== null) {
            (/** @type {?} */ (obj)).slowResource['endTime'] = -1;
            (/** @type {?} */ (obj)).slowResource['name'] = slowestResource.name;
        }
        if (componentTitle !== MezzuriteConstants.altName && componentTitle !== MezzuriteConstants.vltName && renderStartEntry) {
            (/** @type {?} */ (obj))['renderStartTime'] = renderStartEntry.startTime;
        }
        (/** @type {?} */ (window)).mezzurite.measures.push(obj);
    };
    /**
     * Gets measures by name
     * @param name name
     */
    /**
     * Gets measures by name
     * @param {?} name name
     * @return {?}
     */
    PerformanceTimingService.getMeasuresByName = /**
     * Gets measures by name
     * @param {?} name name
     * @return {?}
     */
    function (name) {
        /** @type {?} */
        var result = [];
        if (name === undefined || name === null) {
            return null;
        }
        /** @type {?} */
        var measures = (/** @type {?} */ (window)).mezzurite.measures;
        for (var i = 0; i < measures.length; i++) {
            if (name === measures[i].name) {
                result.push(measures[i]);
            }
        }
        return result;
    };
    /**
 * Gets a specific measure by id
 * @param id id
 */
    /**
     * Gets a specific measure by id
     * @param {?} id id
     * @return {?}
     */
    PerformanceTimingService.getMeasureById = /**
     * Gets a specific measure by id
     * @param {?} id id
     * @return {?}
     */
    function (id) {
        if (id === undefined || id === null) {
            return null;
        }
        /** @type {?} */
        var measures = (/** @type {?} */ (window)).mezzurite.measures;
        for (var i = 0; i < measures.length; i++) {
            if (id === measures[i].id) {
                return measures[i];
            }
        }
        return null;
    };
    /**
* Gets a specific measure by name and id
* @param name name
* @param id id
*/
    /**
     * Gets a specific measure by name and id
     * @param {?} name name
     * @param {?} id id
     * @return {?}
     */
    PerformanceTimingService.getMeasureByNameAndId = /**
     * Gets a specific measure by name and id
     * @param {?} name name
     * @param {?} id id
     * @return {?}
     */
    function (name, id) {
        if (name === undefined || name === null || id === undefined || id === null) {
            return null;
        }
        /** @type {?} */
        var measures = (/** @type {?} */ (window)).mezzurite.measures;
        for (var i = 0; i < measures.length; i++) {
            if (name === measures[i].name && id === measures[i].id) {
                return measures[i];
            }
        }
        return null;
    };
    /**
     * Gets current components from a given capture cycle
     */
    /**
     * Gets current components from a given capture cycle
     * @return {?}
     */
    PerformanceTimingService.getCurrentComponents = /**
     * Gets current components from a given capture cycle
     * @return {?}
     */
    function () {
        /** @type {?} */
        var components = (/** @type {?} */ (window)).mezzurite.measures.filter(function (m) {
            return m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.altName) === -1 &&
                m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.vltName) === -1 &&
                m.startTime >= (/** @type {?} */ (window)).mezzurite.startTime &&
                m.startTime <= (/** @type {?} */ (window)).mezzurite.endTime;
        });
        return components;
    };
    /**
     * Gets lookup object of current components
     */
    /**
     * Gets lookup object of current components
     * @return {?}
     */
    PerformanceTimingService.getCurrentComponentsLookup = /**
     * Gets lookup object of current components
     * @return {?}
     */
    function () {
        /** @type {?} */
        var components = (/** @type {?} */ (window)).mezzurite.measures.filter(function (m) {
            return m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.altName) === -1 &&
                m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.vltName) === -1 &&
                m.startTime >= (/** @type {?} */ (window)).mezzurite.startTime &&
                m.startTime <= (/** @type {?} */ (window)).mezzurite.endTime;
        });
        /** @type {?} */
        var obj = {};
        for (var i = 0; i < components.length; i++) {
            obj[MezzuriteConstants.measureNamePrefix + ';' + components[i].name + ';' + components[i].id] = components[i];
        }
        return obj;
    };
    /**
     * Calculates viewport load time
     */
    /**
     * Calculates viewport load time
     * @return {?}
     */
    PerformanceTimingService.calculateVlt = /**
     * Calculates viewport load time
     * @return {?}
     */
    function () {
        /** @type {?} */
        var maxComponent = null;
        /** @type {?} */
        var maxEndTime = 0;
        /** @type {?} */
        var vltComponents = [];
        /** @type {?} */
        var components = this.getCurrentComponentsLookup();
        /** @type {?} */
        var vltLookup = (/** @type {?} */ (window)).mezzurite.vltComponentLookup;
        /** @type {?} */
        var measure;
        for (var key in vltLookup) {
            if (components[key] && vltLookup[key] === true) {
                vltComponents.push(components[key]);
                if (maxComponent !== null) {
                    /** @type {?} */
                    var slowestResourceEnd = 0;
                    /** @type {?} */
                    var slowestResource = (/** @type {?} */ (window)).mezzurite.slowestResource[key];
                    if (slowestResource !== undefined && slowestResource !== null) {
                        slowestResourceEnd = slowestResource.responseEnd;
                    }
                    /** @type {?} */
                    var maxLast = maxComponent.clt + maxComponent.startTime;
                    /** @type {?} */
                    var currLast = components[key].clt + components[key].startTime;
                    if (currLast > maxLast) {
                        maxComponent = components[key];
                        maxEndTime = currLast;
                    }
                }
                else {
                    maxComponent = components[key];
                }
            }
        }
        if (maxComponent !== null) {
            /** @type {?} */
            var fullName = MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.vltName + ';' + maxComponent.id;
            this.measure(fullName, null, maxComponent.endTime);
            measure = this.getMeasureByNameAndId(MezzuriteConstants.vltName, maxComponent.id);
            performance.clearMarks(MezzuriteConstants.vltMarkStart);
        }
        else {
            return null;
            // no components in view
        }
        return {
            vlt: measure.clt,
            components: vltComponents
        };
    };
    /**
     * Creates sub-element lookup object on global mezzurite object
     * @param el
     * @param key
     */
    /**
     * Creates sub-element lookup object on global mezzurite object
     * @param {?} el
     * @param {?} key
     * @return {?}
     */
    PerformanceTimingService.getElNames = /**
     * Creates sub-element lookup object on global mezzurite object
     * @param {?} el
     * @param {?} key
     * @return {?}
     */
    function (el, key) {
        if ((/** @type {?} */ (window)).mezzurite.childElementNames[key] === undefined) {
            (/** @type {?} */ (window)).mezzurite.childElementNames[key] = [];
        }
        if (el.tagName === 'IMG') {
            (/** @type {?} */ (window)).mezzurite.childElementNames[key].push(el.src);
        }
    };
    /**
     * Calculates slowest resource within a given component element
     * @param el parent element
     * @param fullName component fullname
     */
    /**
     * Calculates slowest resource within a given component element
     * @param {?} el parent element
     * @param {?} fullName component fullname
     * @return {?}
     */
    PerformanceTimingService.calculateSlowestResource = /**
     * Calculates slowest resource within a given component element
     * @param {?} el parent element
     * @param {?} fullName component fullname
     * @return {?}
     */
    function (el, fullName) {
        /** @type {?} */
        var key = MezzuriteUtils.getFullNamePart(fullName, MezzuriteConstants.fullNamePartKey);
        /** @type {?} */
        var slowestResource = null;
        MezzuriteUtils.walkDOM(el, key, this.getElNames);
        /** @type {?} */
        var resources = performance.getEntriesByType('resource').filter(function (r) { return (/** @type {?} */ (r)).initiatorType === 'img'; });
        /** @type {?} */
        var currentResources = (/** @type {?} */ (window)).mezzurite.childElementNames[key];
        if (resources.length === 0) {
            return;
        }
        for (var i = 0; i < currentResources.length; i++) {
            for (var j = 0; j < resources.length; j++) {
                if (currentResources[i] === resources[j].name &&
                    (slowestResource === null || resources[j].responseEnd > slowestResource.responseEnd)) {
                    slowestResource = resources[j];
                }
            }
        }
        (/** @type {?} */ (window)).mezzurite.slowestResource[fullName] = slowestResource;
        return slowestResource;
    };
    /**
     * @return {?}
     */
    PerformanceTimingService.calculateSlowestResourceBatch = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var slow;
        /** @type {?} */
        var elementDict = (/** @type {?} */ (window)).mezzurite.elementLookup;
        for (var prop in elementDict) {
            slow = this.calculateSlowestResource(elementDict[prop], prop);
            if (slow === null) {
                PerformanceTimingService.measure(prop);
            }
            else {
                PerformanceTimingService.measure(prop, slow);
            }
        }
    };
    return PerformanceTimingService;
}());
/**
 * Class containing core timing functions
 */
export { PerformanceTimingService };
