// @microsoft/mezzurite-core/fesm2015/mezzurite-core.js
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** *
 * Global Mezzurite constants
  @type {?} */
const MezzuriteConstants = {
    mezzuriteObjectVersion: '1.0.0',
    captureCycleTimeout: 10000,
    slowestResourceTimeout: 4000,
    idLength: 6,
    domAttributeName: 'perf-id',
    measureNamePrefix: 'mz',
    altName: 'ALT',
    vltName: 'VLT',
    fvltName: 'FVLT',
    vltMarkStart: 'VltStart',
    altMarkStart: 'AltStart',
    altMarkEnd: 'AltEnd',
    componentMarkStart: 'ComponentStart',
    componentMarkEnd: 'ComponentEnd',
    componentMarkRenderStart: 'ComponentRenderStart',
    jsllConfigName: 'jsll',
    versionName: 'MezzuriteVersion',
    allComponents: 'AllComponents',
    redirect: 'Redirect',
    sessionData: 'MezzuriteSession',
    fullNamePartTitle: 'title',
    fullNamePartKey: 'key',
    unsupportedBrowserName: 'unsupportedBrowser',
    unsupportedBrowserPerf: 'This was sent from a client using a browser that does not support the Performance API'
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Global Mezzurite object
 */
class MezzuriteObject {
    constructor() {
        this.firstViewLoaded = false;
        this.captureCycleStarted = false;
        this.routerPerf = false;
        this.measures = [];
        this.defaultLogs = [];
        this.childElementNames = {};
        this.slowestResource = {};
        this.currentComponents = {};
        this.vltComponentLookup = {};
        this.elementLookup = {};
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Class of utility functions for Mezzurite
 */
class MezzuriteUtils {
    constructor() { }
    /**
     * Adds default values for MezzuriteObject to the global window.mezzurite object
     * @param {?} obj existing mezzurite global object (if already created by logger)
     * @return {?}
     */
    static createMezzuriteObject(obj) {
        this.addCustomEventPolyfill();
        /** @type {?} */
        const mzObj = new MezzuriteObject();
        for (const prop in mzObj) {
            if (obj[prop] === undefined) {
                obj[prop] = (/** @type {?} */ (mzObj))[prop];
            }
        }
    }
    /**
     * Resets certain properties in window.mezzurite after capture cycle is completed
     * @return {?}
     */
    static testReset() {
        /** @type {?} */
        const obj = (/** @type {?} */ (window)).mezzurite;
        obj.childElementNames = {};
        obj.slowestResource = {};
        obj.currentComponents = {};
        obj.vltComponentLookup = {};
        obj.elementLookup = {};
        (/** @type {?} */ (window)).mezzurite = obj;
    }
    /**
     * Creates a unique alpha-numeric key
     * @return {?}
     */
    static makeId() {
        /** @type {?} */
        let text = '';
        /** @type {?} */
        const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        for (let i = 0; i < MezzuriteConstants.idLength; i++) {
            text += possible.charAt(Math.floor(Math.random() * possible.length));
        }
        return text;
    }
    /**
     * Gets string name of the given function
     * @param {?} fun function
     * @return {?}
     */
    static getFunctionName(fun) {
        /** @type {?} */
        let ret = fun.toString();
        ret = ret.substr('function '.length);
        ret = ret.substr(0, ret.indexOf('('));
        return ret;
    }
    /**
     * Gets name of a given component passed into Mezzurite HOC
     * @param {?} WrappedComponent Component
     * @return {?}
     */
    static getDisplayName(WrappedComponent) {
        if (WrappedComponent.name !== undefined) {
            return WrappedComponent.name;
        }
        return MezzuriteUtils.getFunctionName(WrappedComponent);
    }
    /**
     * Gets complete name consisting of prefix, component name, and unique key.
     * @param {?} name component name
     * @param {?} key unique key
     * @return {?}
     */
    static getName(name, key) {
        return MezzuriteConstants.measureNamePrefix + ';' + name + ';' + key;
    }
    /**
     * Creates metric to save to global mezzurite object
     * @param {?} metricType metric type
     * @param {?} value measured value
     * @param {?=} data json metadata
     * @return {?}
     */
    static createMetric(metricType, value, data = null) {
        /** @type {?} */
        const obj = {
            metricType: metricType,
            value: value % 1 !== 0 ? parseFloat(value.toFixed(1)) : value
        };
        if (data !== null) {
            obj.data = JSON.stringify(data);
        }
        return obj;
    }
    /**
     * Walks DOM of a given element
     * @param {?} node DOM node
     * @param {?} key unique identifier
     * @param {?} func callback function
     * @return {?}
     */
    static walkDOM(node, key, func) {
        func(node, key);
        node = node.firstChild;
        while (node) {
            MezzuriteUtils.walkDOM(node, key, func);
            node = node.nextSibling;
        }
    }
    /**
     * Gets fragment from full name
     * @param {?} fullName full name
     * @param {?} val value to pull
     * @return {?}
     */
    static getFullNamePart(fullName, val) {
        /** @type {?} */
        const arr = fullName.split(';');
        switch (val) {
            case MezzuriteConstants.fullNamePartTitle:
                return arr[1];
            case MezzuriteConstants.fullNamePartKey:
                return arr[2];
            default:
                return fullName;
        }
    }
    /**
     * Polyfill that adds CustomEvent for IE usage
     * @return {?}
     */
    static addCustomEventPolyfill() {
        if (typeof (/** @type {?} */ (window)).CustomEvent === 'function') {
            return false;
        }
        /**
         * @param {?} event
         * @param {?} params
         * @return {?}
         */
        function CustomEvent(event, params) {
            params = params || { bubbles: false, cancelable: false, detail: undefined };
            /** @type {?} */
            const evt = document.createEvent('CustomEvent');
            evt.initCustomEvent(event, params.bubbles, params.cancelable, params.detail);
            return evt;
        }
        CustomEvent.prototype = (/** @type {?} */ (window)).Event.prototype;
        (/** @type {?} */ (window)).CustomEvent = CustomEvent;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Mezzurite Measure
 */
class MezzuriteMeasure {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Class containing core timing functions
 */
class PerformanceTimingService {
    /**
     * Creates measure object from given set of performance marks
     * @param {?} name full semicolon delimited name
     * @param {?=} slowestResource Slowest resource inside component
     * @param {?=} maxComponentEndTime Max component end time (if VLT)
     * @return {?}
     */
    static measure(name, slowestResource = null, maxComponentEndTime = null) {
        /** @type {?} */
        let startEntry;
        /** @type {?} */
        let endEntry;
        /** @type {?} */
        const componentTitle = MezzuriteUtils.getFullNamePart(name, MezzuriteConstants.fullNamePartTitle);
        /** @type {?} */
        const key = MezzuriteUtils.getFullNamePart(name, MezzuriteConstants.fullNamePartKey);
        if (name === undefined) {
            return;
        }
        if (componentTitle === MezzuriteConstants.altName) {
            // ALT
            endEntry = performance.getEntriesByName(MezzuriteConstants.altMarkEnd)[0];
        }
        else if (componentTitle === MezzuriteConstants.vltName) {
            // VLT
            startEntry = performance.getEntriesByName(MezzuriteConstants.vltMarkStart)[0];
            endEntry = {
                startTime: maxComponentEndTime
            };
        }
        else {
            // Component
            startEntry = performance.getEntriesByName(key + MezzuriteConstants.componentMarkStart)[0];
            endEntry = performance.getEntriesByName(key + MezzuriteConstants.componentMarkEnd)[0];
        }
        /** @type {?} */
        const renderStartEntry = performance.getEntriesByName(key + MezzuriteConstants.componentMarkRenderStart)[0];
        /** @type {?} */
        const startTime = startEntry !== undefined ? startEntry.startTime : 0;
        /** @type {?} */
        let endTime = endEntry.startTime;
        // reset end time to end of slowest resource if
        if (slowestResource !== null && slowestResource.responseEnd > endTime) {
            endTime = slowestResource.responseEnd;
        }
        /** @type {?} */
        const mountDuration = endEntry.startTime - startTime;
        /** @type {?} */
        const totalDuration = endTime - startTime;
        /** @type {?} */
        const nameArr = name.split(';');
        /** @type {?} */
        const obj = new MezzuriteMeasure();
        obj.name = nameArr[1];
        obj.id = nameArr[2];
        obj.startTime = startTime % 1 !== 0 ? parseFloat(startTime.toFixed(1)) : startTime;
        obj.endTime = parseFloat(endTime.toFixed(1));
        obj.untilMount = parseFloat(mountDuration.toFixed(1));
        obj.clt = parseFloat(totalDuration.toFixed(1));
        obj.slowResource = {};
        if (slowestResource && slowestResource.responseEnd >= startTime) {
            (/** @type {?} */ (obj)).slowResource['endTime'] = parseFloat(slowestResource.responseEnd.toFixed(1));
            (/** @type {?} */ (obj)).slowResource['name'] = slowestResource.name;
        }
        else if (slowestResource !== null) {
            (/** @type {?} */ (obj)).slowResource['endTime'] = -1;
            (/** @type {?} */ (obj)).slowResource['name'] = slowestResource.name;
        }
        if (componentTitle !== MezzuriteConstants.altName && componentTitle !== MezzuriteConstants.vltName && renderStartEntry) {
            (/** @type {?} */ (obj))['renderStartTime'] = renderStartEntry.startTime;
        }
        (/** @type {?} */ (window)).mezzurite.measures.push(obj);
    }
    /**
     * Gets measures by name
     * @param {?} name name
     * @return {?}
     */
    static getMeasuresByName(name) {
        /** @type {?} */
        const result = [];
        if (name === undefined || name === null) {
            return null;
        }
        /** @type {?} */
        const measures = (/** @type {?} */ (window)).mezzurite.measures;
        for (let i = 0; i < measures.length; i++) {
            if (name === measures[i].name) {
                result.push(measures[i]);
            }
        }
        return result;
    }
    /**
     * Gets a specific measure by id
     * @param {?} id id
     * @return {?}
     */
    static getMeasureById(id) {
        if (id === undefined || id === null) {
            return null;
        }
        /** @type {?} */
        const measures = (/** @type {?} */ (window)).mezzurite.measures;
        for (let i = 0; i < measures.length; i++) {
            if (id === measures[i].id) {
                return measures[i];
            }
        }
        return null;
    }
    /**
     * Gets a specific measure by name and id
     * @param {?} name name
     * @param {?} id id
     * @return {?}
     */
    static getMeasureByNameAndId(name, id) {
        if (name === undefined || name === null || id === undefined || id === null) {
            return null;
        }
        /** @type {?} */
        const measures = (/** @type {?} */ (window)).mezzurite.measures;
        for (let i = 0; i < measures.length; i++) {
            if (name === measures[i].name && id === measures[i].id) {
                return measures[i];
            }
        }
        return null;
    }
    /**
     * Gets current components from a given capture cycle
     * @return {?}
     */
    static getCurrentComponents() {
        /** @type {?} */
        const components = (/** @type {?} */ (window)).mezzurite.measures.filter((m) => m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.altName) === -1 &&
            m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.vltName) === -1 &&
            m.startTime >= (/** @type {?} */ (window)).mezzurite.startTime &&
            m.startTime <= (/** @type {?} */ (window)).mezzurite.endTime);
        return components;
    }
    /**
     * Gets lookup object of current components
     * @return {?}
     */
    static getCurrentComponentsLookup() {
        /** @type {?} */
        const components = (/** @type {?} */ (window)).mezzurite.measures.filter((m) => m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.altName) === -1 &&
            m.name.indexOf(MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.vltName) === -1 &&
            m.startTime >= (/** @type {?} */ (window)).mezzurite.startTime &&
            m.startTime <= (/** @type {?} */ (window)).mezzurite.endTime);
        /** @type {?} */
        const obj = {};
        for (let i = 0; i < components.length; i++) {
            obj[MezzuriteConstants.measureNamePrefix + ';' + components[i].name + ';' + components[i].id] = components[i];
        }
        return obj;
    }
    /**
     * Calculates viewport load time
     * @return {?}
     */
    static calculateVlt() {
        /** @type {?} */
        let maxComponent = null;
        /** @type {?} */
        const vltComponents = [];
        /** @type {?} */
        const components = this.getCurrentComponentsLookup();
        /** @type {?} */
        const vltLookup = (/** @type {?} */ (window)).mezzurite.vltComponentLookup;
        /** @type {?} */
        let measure;
        for (const key in vltLookup) {
            if (components[key] && vltLookup[key] === true) {
                vltComponents.push(components[key]);
                if (maxComponent !== null) {
                    /** @type {?} */
                    let slowestResourceEnd = 0;
                    /** @type {?} */
                    const slowestResource = (/** @type {?} */ (window)).mezzurite.slowestResource[key];
                    if (slowestResource !== undefined && slowestResource !== null) {
                        slowestResourceEnd = slowestResource.responseEnd;
                    }
                    /** @type {?} */
                    const maxLast = maxComponent.clt + maxComponent.startTime;
                    /** @type {?} */
                    const currLast = components[key].clt + components[key].startTime;
                    if (currLast > maxLast) {
                        maxComponent = components[key];
                    }
                }
                else {
                    maxComponent = components[key];
                }
            }
        }
        if (maxComponent !== null) {
            /** @type {?} */
            const fullName = MezzuriteConstants.measureNamePrefix + ';' + MezzuriteConstants.vltName + ';' + maxComponent.id;
            this.measure(fullName, null, maxComponent.endTime);
            measure = this.getMeasureByNameAndId(MezzuriteConstants.vltName, maxComponent.id);
            performance.clearMarks(MezzuriteConstants.vltMarkStart);
        }
        else {
            return null;
            // no components in view
        }
        return {
            vlt: measure.clt,
            components: vltComponents
        };
    }
    /**
     * Creates sub-element lookup object on global mezzurite object
     * @param {?} el
     * @param {?} key
     * @return {?}
     */
    static getElNames(el, key) {
        if ((/** @type {?} */ (window)).mezzurite.childElementNames[key] === undefined) {
            (/** @type {?} */ (window)).mezzurite.childElementNames[key] = [];
        }
        if (el.tagName === 'IMG') {
            (/** @type {?} */ (window)).mezzurite.childElementNames[key].push(el.src);
        }
    }
    /**
     * Calculates slowest resource within a given component element
     * @param {?} el parent element
     * @param {?} fullName component fullname
     * @return {?}
     */
    static calculateSlowestResource(el, fullName) {
        /** @type {?} */
        const key = MezzuriteUtils.getFullNamePart(fullName, MezzuriteConstants.fullNamePartKey);
        /** @type {?} */
        let slowestResource = null;
        MezzuriteUtils.walkDOM(el, key, this.getElNames);
        /** @type {?} */
        const resources = performance.getEntriesByType('resource').filter((r) => (/** @type {?} */ (r)).initiatorType === 'img');
        /** @type {?} */
        const currentResources = (/** @type {?} */ (window)).mezzurite.childElementNames[key];
        if (resources.length === 0) {
            return;
        }
        for (let i = 0; i < currentResources.length; i++) {
            for (let j = 0; j < resources.length; j++) {
                if (currentResources[i] === resources[j].name &&
                    (slowestResource === null || resources[j].responseEnd > slowestResource.responseEnd)) {
                    slowestResource = resources[j];
                }
            }
        }
        (/** @type {?} */ (window)).mezzurite.slowestResource[fullName] = slowestResource;
        return slowestResource;
    }
    /**
     * @return {?}
     */
    static calculateSlowestResourceBatch() {
        /** @type {?} */
        let slow;
        /** @type {?} */
        const elementDict = (/** @type {?} */ (window)).mezzurite.elementLookup;
        for (const prop in elementDict) {
            slow = this.calculateSlowestResource(elementDict[prop], prop);
            if (slow === null) {
                PerformanceTimingService.measure(prop);
            }
            else {
                PerformanceTimingService.measure(prop, slow);
            }
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Class containing core telemetry functions
 */
class PerformanceTelemetryService {
    /**
     * Starts capture cycle period
     * @return {?}
     */
    static startCaptureCycle() {
        if (!(/** @type {?} */ (window)).mezzurite.captureCycleStarted) {
            (/** @type {?} */ (window)).mezzurite.startTime = window.performance.now();
            (/** @type {?} */ (window)).mezzurite.captureCycleStarted = true;
            (/** @type {?} */ (window)).mezzurite.captureTimer = setTimeout(function () {
                PerformanceTelemetryService.captureTimings();
            }, MezzuriteConstants.captureCycleTimeout);
        }
    }
    /**
     * Captures timings for the given period
     * @param {?=} isRedirect Bool dictating whether timings were captured at end of cycle or early
     * @return {?}
     */
    static captureTimings(isRedirect = false) {
        clearTimeout((/** @type {?} */ (window)).mezzurite.captureTimer);
        (/** @type {?} */ (window)).mezzurite.endTime = window.performance.now();
        if (!(/** @type {?} */ (window)).mezzurite.captureCycleStarted) {
            (/** @type {?} */ (window)).mezzurite.captureCycleStarted = true;
        }
        PerformanceTelemetryService.submitTelemetry(isRedirect);
        (/** @type {?} */ (window)).mezzurite.captureCycleStarted = false;
    }
    /**
     * Creates timings object to send to telemetry
     * @param {?} isRedirect isRedirect bool
     * @return {?}
     */
    static submitTelemetry(isRedirect) {
        /** @type {?} */
        const timings = [];
        // add redirect value
        timings.push(MezzuriteUtils.createMetric(MezzuriteConstants.redirect, isRedirect === false ? 0 : 1));
        // calculate component measures off slowest resource values
        if ((/** @type {?} */ (window)).mezzurite.elementLookup !== {}) {
            PerformanceTimingService.calculateSlowestResourceBatch();
        }
        /** @type {?} */
        const components = PerformanceTimingService.getCurrentComponents();
        if ((/** @type {?} */ (window)).mezzurite.routerPerf) {
            // alt
            if ((/** @type {?} */ (window)).mezzurite.firstViewLoaded === false) {
                /** @type {?} */
                const altMeasure = (/** @type {?} */ (window)).mezzurite.measures.filter((m) => m.name.indexOf(MezzuriteConstants.altName) > -1)[0];
                timings.push(MezzuriteUtils.createMetric(MezzuriteConstants.altName, altMeasure.clt));
                (/** @type {?} */ (window)).mezzurite.firstViewLoaded = true;
            }
            // vlt
            if (components.length > 0) {
                /** @type {?} */
                const vltResults = PerformanceTimingService.calculateVlt();
                if (vltResults !== null) {
                    timings.push(MezzuriteUtils.createMetric(MezzuriteConstants.vltName, vltResults.vlt, vltResults.components));
                }
            }
            if (components.length === 0) {
                performance.clearMarks(MezzuriteConstants.vltMarkStart);
            }
        }
        if (components.length > 0) {
            timings.push(MezzuriteUtils.createMetric(MezzuriteConstants.allComponents, -1, components));
        }
        this.log(timings);
        MezzuriteUtils.testReset();
    }
    /**
     * Adds remaining metadata to send to logger and dispatches event
     * @param {?} timings
     * @return {?}
     */
    static log(timings) {
        if ((/** @type {?} */ (window)).mezzurite) {
            if (timings.length > 1) {
                /** @type {?} */
                const obj = {
                    Timings: timings,
                    Framework: {
                        name: (/** @type {?} */ (window)).mezzurite.packageName,
                        version: (/** @type {?} */ (window)).mezzurite.packageVersion
                    },
                    ViewportWidth: (/** @type {?} */ (window)).mezzurite.viewportWidth,
                    ViewportHeight: (/** @type {?} */ (window)).mezzurite.viewportHeight,
                    ObjectVersion: MezzuriteConstants.mezzuriteObjectVersion
                };
                // log to console when developing locally
                if ((/** @type {?} */ (window)).location.href.indexOf('localhost') > -1) {
                    console.log('to log for testing: ', obj);
                }
                if ((/** @type {?} */ (window)).mezzurite.EventElement) {
                    (/** @type {?} */ (window)).mezzurite.EventElement.dispatchEvent(new CustomEvent('Timing', { detail: obj }));
                }
            }
            else {
                console.log('nothing for Mezzurite to log.');
            }
        }
    }
    /**
     * Checks whether window.performance is undefined
     * @return {?}
     */
    static compatibilityCheck() {
        /** @type {?} */
        const isCompatible = (window.performance !== undefined);
        if (!isCompatible) {
            /** @type {?} */
            const timings = [
                MezzuriteUtils.createMetric(MezzuriteConstants.unsupportedBrowserName, -1, MezzuriteConstants.unsupportedBrowserPerf)
            ];
            this.log(timings);
        }
        return isCompatible;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Class for a given performance metric
 */
class PerfMetric {
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// This file only reexports content of the `src` folder. Keep it that way.

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { PerformanceTelemetryService, PerformanceTimingService, PerfMetric, MezzuriteConstants, MezzuriteUtils };
