// @microsoft/mezzurite-angular/fesm2015/mezzurite-angular.js
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
import { MezzuriteUtils, PerformanceTimingService, PerformanceTelemetryService, MezzuriteConstants } from '@microsoft/mezzurite-core';
import { filter } from 'rxjs/operators';
import { Injectable, Directive, ElementRef, Input, NgModule } from '@angular/core';
import { Router, NavigationStart } from '@angular/router';
import 'intersection-observer';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** *
 * Gets information from package.json
  @type {?} */
const environment = {
    version: require('../package.json').version,
    name: require('../package.json').name
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Extension of Mezzurite Utilities that gets package specific information
 */
class MezzuriteAngularUtils extends MezzuriteUtils {
    /**
     * @param {?} obj
     * @return {?}
     */
    static createMezzuriteObject(obj) {
        super.createMezzuriteObject(obj);
        (/** @type {?} */ (window)).mezzurite.packageVersion = environment.version;
        (/** @type {?} */ (window)).mezzurite.packageName = environment.name;
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * RoutingService is responsible for listening to the routing events coming
 * into the application and creating performance marks based on them.
 */
class RoutingService {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        this.endCounter = 0;
        if (!(/** @type {?} */ (window)).mezzurite) {
            (/** @type {?} */ (window)).mezzurite = {};
        }
        MezzuriteAngularUtils.createMezzuriteObject((/** @type {?} */ (window)).mezzurite);
        (/** @type {?} */ (window)).mezzurite.routerPerf = true;
    }
    /**
     * this method begins the listening process. Must be called for code to function properly.
     * @return {?}
     */
    start() {
        /** @type {?} */
        const onNavStart$ = (/** @type {?} */ (this)).router.events.pipe(filter(event => event instanceof NavigationStart));
        onNavStart$.subscribe((e) => {
            this.handleRoute(e);
        });
    }
    /**
     * @param {?} e
     * @return {?}
     */
    handleRoute(e) {
        (/** @type {?} */ (window)).mezzurite.routeUrl = e.url;
        if ((/** @type {?} */ (window)).mezzurite.captureCycleStarted) {
            (/** @type {?} */ (window)).mezzurite.captureCycleStarted = false;
            PerformanceTelemetryService.captureTimings(true);
            window.performance.mark(MezzuriteConstants.vltMarkStart);
            // starts a new capture cycle
            PerformanceTelemetryService.startCaptureCycle();
        }
        else {
            // starts the capture cycle to transmit telemetry if current pathname is different than recentPath
            PerformanceTelemetryService.startCaptureCycle();
            // If first load, capture ALT
            if (!(/** @type {?} */ (window)).mezzurite.firstViewLoaded) {
                window.performance.mark(MezzuriteConstants.altMarkEnd);
                window.performance.mark(MezzuriteConstants.vltMarkStart);
                /** @type {?} */
                const fullName = MezzuriteAngularUtils.getName(MezzuriteConstants.altName, MezzuriteAngularUtils.makeId());
                PerformanceTimingService.measure(fullName, 0, MezzuriteConstants.altMarkEnd);
            }
            else {
                window.performance.mark(MezzuriteConstants.vltMarkStart);
            }
        }
    }
}
RoutingService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
RoutingService.ctorParameters = () => [
    { type: Router }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Directive that enables tracking on specific component elements
 */
class MezzuriteDirective {
    /**
     * @param {?} ref
     */
    constructor(ref) {
        // tslint:disable-next-line:no-input-rename
        this.title = 'MyComponent';
        this.id = MezzuriteUtils.makeId();
        performance.mark(this.id + MezzuriteConstants.componentMarkStart);
        this.el = ref.nativeElement;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.fullName = MezzuriteUtils.getName(this.title, this.id);
        (/** @type {?} */ (window)).mezzurite.elementLookup[this.fullName] = this.el;
        /** @type {?} */
        const that = this;
        /** @type {?} */
        const config = {
            root: /** @type {?} */ (null),
            // setting it to 'null' sets it to default value: viewport
            rootMargin: '0px'
        };
        /** @type {?} */
        const intObserver = new IntersectionObserver(function (entries, observer) {
            performance.mark(that.id + MezzuriteConstants.componentMarkEnd);
            /** @type {?} */
            const entry = entries[0];
            (/** @type {?} */ (window)).mezzurite.viewportWidth = entry.rootBounds.width;
            (/** @type {?} */ (window)).mezzurite.viewportHeight = entry.rootBounds.height;
            if (entry.isIntersecting) {
                (/** @type {?} */ (window)).mezzurite.vltComponentLookup[that.fullName] = true;
            }
            observer.unobserve(that.el);
        }, config);
        intObserver.observe(this.el);
    }
}
MezzuriteDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mezzurite]'
            },] }
];
/** @nocollapse */
MezzuriteDirective.ctorParameters = () => [
    { type: ElementRef }
];
MezzuriteDirective.propDecorators = {
    title: [{ type: Input, args: ['component-title',] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Mezzurite Performance Module for Angular
 */
class AngularPerfModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: AngularPerfModule,
            providers: [
                RoutingService
            ]
        };
    }
}
AngularPerfModule.decorators = [
    { type: NgModule, args: [{
                declarations: [MezzuriteDirective],
                exports: [MezzuriteDirective]
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// This file only reexports content of the `src` folder. Keep it that way.

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { RoutingService, AngularPerfModule, MezzuriteDirective as ɵa };
