// @microsoft/mezzurite-core/esm2015/src/utils/performance-constants.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** *
 * Global Mezzurite constants
  @type {?} */
const MezzuriteConstants = {
    mezzuriteObjectVersion: '1.0.0',
    captureCycleTimeout: 10000,
    slowestResourceTimeout: 4000,
    idLength: 6,
    domAttributeName: 'perf-id',
    measureNamePrefix: 'mz',
    altName: 'ALT',
    vltName: 'VLT',
    fvltName: 'FVLT',
    vltMarkStart: 'VltStart',
    altMarkStart: 'AltStart',
    altMarkEnd: 'AltEnd',
    componentMarkStart: 'ComponentStart',
    componentMarkEnd: 'ComponentEnd',
    componentMarkRenderStart: 'ComponentRenderStart',
    jsllConfigName: 'jsll',
    versionName: 'MezzuriteVersion',
    allComponents: 'AllComponents',
    redirect: 'Redirect',
    sessionData: 'MezzuriteSession',
    fullNamePartTitle: 'title',
    fullNamePartKey: 'key',
    unsupportedBrowserName: 'unsupportedBrowser',
    unsupportedBrowserPerf: 'This was sent from a client using a browser that does not support the Performance API'
};
export { MezzuriteConstants };
