// @microsoft/mezzurite-core/esm2015/src/services/performance-utils.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { MezzuriteObject } from '../utils/performance-global';
import { MezzuriteConstants } from '../utils/performance-constants';
/**
 * Class of utility functions for Mezzurite
 */
export class MezzuriteUtils {
    constructor() { }
    /**
     * Adds default values for MezzuriteObject to the global window.mezzurite object
     * @param {?} obj existing mezzurite global object (if already created by logger)
     * @return {?}
     */
    static createMezzuriteObject(obj) {
        this.addCustomEventPolyfill();
        /** @type {?} */
        const mzObj = new MezzuriteObject();
        for (const prop in mzObj) {
            if (obj[prop] === undefined) {
                obj[prop] = (/** @type {?} */ (mzObj))[prop];
            }
        }
    }
    /**
     * Resets certain properties in window.mezzurite after capture cycle is completed
     * @return {?}
     */
    static testReset() {
        /** @type {?} */
        const obj = (/** @type {?} */ (window)).mezzurite;
        obj.childElementNames = {};
        obj.slowestResource = {};
        obj.currentComponents = {};
        obj.vltComponentLookup = {};
        obj.elementLookup = {};
        (/** @type {?} */ (window)).mezzurite = obj;
    }
    /**
     * Creates a unique alpha-numeric key
     * @return {?}
     */
    static makeId() {
        /** @type {?} */
        let text = '';
        /** @type {?} */
        const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        for (let i = 0; i < MezzuriteConstants.idLength; i++) {
            text += possible.charAt(Math.floor(Math.random() * possible.length));
        }
        return text;
    }
    /**
     * Gets string name of the given function
     * @param {?} fun function
     * @return {?}
     */
    static getFunctionName(fun) {
        /** @type {?} */
        let ret = fun.toString();
        ret = ret.substr('function '.length);
        ret = ret.substr(0, ret.indexOf('('));
        return ret;
    }
    /**
     * Gets name of a given component passed into Mezzurite HOC
     * @param {?} WrappedComponent Component
     * @return {?}
     */
    static getDisplayName(WrappedComponent) {
        if (WrappedComponent.name !== undefined) {
            return WrappedComponent.name;
        }
        return MezzuriteUtils.getFunctionName(WrappedComponent);
    }
    /**
     * Gets complete name consisting of prefix, component name, and unique key.
     * @param {?} name component name
     * @param {?} key unique key
     * @return {?}
     */
    static getName(name, key) {
        return MezzuriteConstants.measureNamePrefix + ';' + name + ';' + key;
    }
    /**
     * Creates metric to save to global mezzurite object
     * @param {?} metricType metric type
     * @param {?} value measured value
     * @param {?=} data json metadata
     * @return {?}
     */
    static createMetric(metricType, value, data = null) {
        /** @type {?} */
        const obj = {
            metricType: metricType,
            value: value % 1 !== 0 ? parseFloat(value.toFixed(1)) : value
        };
        if (data !== null) {
            obj.data = JSON.stringify(data);
        }
        return obj;
    }
    /**
     * Walks DOM of a given element
     * @param {?} node DOM node
     * @param {?} key unique identifier
     * @param {?} func callback function
     * @return {?}
     */
    static walkDOM(node, key, func) {
        func(node, key);
        node = node.firstChild;
        while (node) {
            MezzuriteUtils.walkDOM(node, key, func);
            node = node.nextSibling;
        }
    }
    /**
     * Gets fragment from full name
     * @param {?} fullName full name
     * @param {?} val value to pull
     * @return {?}
     */
    static getFullNamePart(fullName, val) {
        /** @type {?} */
        const arr = fullName.split(';');
        switch (val) {
            case MezzuriteConstants.fullNamePartTitle:
                return arr[1];
            case MezzuriteConstants.fullNamePartKey:
                return arr[2];
            default:
                return fullName;
        }
    }
    /**
     * Polyfill that adds CustomEvent for IE usage
     * @return {?}
     */
    static addCustomEventPolyfill() {
        if (typeof (/** @type {?} */ (window)).CustomEvent === 'function') {
            return false;
        }
        /**
         * @param {?} event
         * @param {?} params
         * @return {?}
         */
        function CustomEvent(event, params) {
            params = params || { bubbles: false, cancelable: false, detail: undefined };
            /** @type {?} */
            const evt = document.createEvent('CustomEvent');
            evt.initCustomEvent(event, params.bubbles, params.cancelable, params.detail);
            return evt;
        }
        CustomEvent.prototype = (/** @type {?} */ (window)).Event.prototype;
        (/** @type {?} */ (window)).CustomEvent = CustomEvent;
    }
}
